### PR TITLE
[ENH] Add null_flag function #501

### LINKED ISSUE #501: [ENH] Add ability to flag null values in columns
# Brief Description

Hello everyone, I'd like to suggest a new method (I'll tentatively call it `flag_nulls`) that adds a new column to the dataframe to indicate if there are null values in the row.

If you are preparing a dataframe for machine learning or something, it's important to fill in the null values with something. However, the fact that the data is null could, in fact, be its own feature--for example, if someone is submitting an insurance claim and they haven't provided important information about themselves, that might be a flag that the claim is fraudulent.

# Example API

```python
# flag null values in columns
df.flag_nulls(column_name='null_flag', columns=None)
# column_name gives the name of the new column we generate
# columns is a list of the columns that we check for null values. If columns is None, we use all the columns

df1 = pd.DataFrame({'a': [1, 2, None, 4], 'b': [5.0, None, 7.0, 8.0]}

print df1.flag_nulls()
```
a | b | null_flag
---|---|---
1    | 5.0 | 0
2   | None | 1
None | 7.0 | 1
4 | 8.0 | 0

```python
print df1.flag_nulls(columns=['a'])
```
a | b | null_flag
---|---|---
1    | 5.0 | 0
2   | None | 0
None | 7.0 | 1
4 | 8.0 | 0

```python
print df1.flag_nulls(columns=['a'], column_name='flag')
```
a | b | flag
---|---|---
1    | 5.0 | 0
2   | None | 0
None | 7.0 | 1
4 | 8.0 | 0

# Notes

- Computerphile [agrees with me](https://www.youtube.com/watch?v=oCQbC818KKU) on this strategy as well ;)
- I would like to work on this issue if that's acceptable.

### PR DESCRIPTION
**Pull request by anzelpwj**
# PR Description

Adds a null flag function, which creates a new column in a DataFrame to mark which rows had null values. You can choose a subset of columns to consider as needed.

**This PR resolves #501.**

# PR Checklist

Please ensure that you have done the following:

1. [X] PR in from a fork off your branch. Do not PR from `<your_username>`:master, but rather from `<your_username>`:<branch_name>.
2. [X] If you're not on the contributors list, add yourself to `AUTHORS.rst`.
3. [X] Add a line to `CHANGELOG.rst` under the latest version header (i.e. the one that is "on deck") describing the contribution.

## Quick Check

To do a very quick check that everything is correct, follow these steps below:

- [X] Run the command `make check` from pyjanitor's top-level directory. This will automatically run:
    - black formatting
    - pycodestyle checking
    - running the test suite
    - docs build
    
Once done, please check off the check-box above.
    
If `make check` does not work for you, you can execute the commands listed in the Makefile individually.

## Code Changes

If you are adding code changes, please ensure the following:

- [X] Ensure that you have added tests.
- [X] Run all tests (`$ pytest .`) locally on your machine.
    - [X] Check to ensure that test coverage covers the lines of code that you have added.
    - [X] Ensure that all tests pass.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl
- @jk3587

### CODE REVIEW

**Comment by anzelpwj** [comment]
Just realized I probably need to add some files to the docs folder. Will do that tomorrow.

File: tests/functions/test_flag_nulls.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import pandas as pd
+import pytest
+from pandas.testing import assert_frame_equal
+
+TEST_DF = pd.DataFrame({"a": [1, 2, None, 4], "b": [5.0, None, 7.0, 8.0]})
```
Comment (ericmjl) [inline tests/functions/test_flag_nulls.py]: Wondering if you might be open to putting this in `conftest.py` as a fixture? 

We also have a `nulldf` fixture inside `conftest.py`, which may come in handy, and reduce code duplication.

File: janitor/functions.py [hunk 2]
```diff
@@ -3309,3 +3309,74 @@ def take_a_if_even(x):
...
+    df: pd.DataFrame,
+    column_name: str = "null_flag",
+    columns: Union[Iterable[str], str, None] = None,
+) -> pd.DataFrame:
+    """
+    Creates a new column to indicate whether you have null values in a given
+    row. If the columns parameter is not set, looks across the entire
+    DataFrame, otherwise will look only in the columns you set.
+
+    .. code-block:: python
```
Comment (ericmjl) [inline janitor/functions.py]: We recently started standardizing docstrings (h/t @loganthomas for kickstarting this). For pragmatic reasons (it’d be a lot of functions to cover), we haven’t put in toy examples in the docstrings, but I think in the long-run it’s a better idea to do what you’ve done. For now, we just have the “Functional” and “Method Chaining” examples, without inputs and outputs. 

Having written all that, let’s stick with what you’ve done here. I’ll make an issue to ask for the examples in the docstrings to be changed to minimal examples like you’ve provided, but nonetheless showcasing both the functional and method chaining styles.

**Comment by ericmjl** [comment]
Everything looks great, thanks @anzelpwj!

**Comment by hectormz** [comment]
@ericmjl, @anzelpwj  I can move this conversation to an issue if more appropriate.

I rebased recently and included this PR. I'm getting some failing tests that were added with this PR:
* test_non_method_functional
* test_functional_on_some_columns
* test_rename_output_column

The issue is that `flag_nulls()` is adding a column with `dtype=int32`, but the `expected` dataframe is adding a column that is `int64`. Not sure why that might only be happening for me, and passed tests for all of you and Azure.

Adding a line like:
```python
expected = expected.astype({"null_flag": "int32"})
```
fixes the issues. Any thoughts?

I'm on Windows in a pyjanitor venv

**Comment by anzelpwj** [comment]
Hmmm, what if we are explicitly setting `check_column_type='equiv'` in the dataframes assert? Does that fix it?

**Comment by hectormz** [comment]
Hmm, I believe that is the default, and setting it explicitly didn't change for me. So that's even stranger that this is happening.

**Comment by anzelpwj** [comment]
It is default, but thought we'd check anyway. I'd be fine with this fix, but TBH we should open up a ticket on the Pandas project to list this as a bug with their assert method.

**Comment by Ram-N** [comment]
I am on Windows10 and getting the exact same 4 tests failing, reported by @hectormz above. 

```
 assert_frame_equal(df, expected)
E       AssertionError: Attributes are different
E       
E       Attribute "dtype" are different
E       [left]:  int32
E       [right]: int64
```
The thing is, I am just getting the dev env set up. I haven't made a single change to the code or to the docs yet. 

Is there any way to 'soften' these tests so that int32 and int64 are treated as equivalent dtypes?

**Comment by hectormz** [comment]
@anzelpwj it might be an issue with Pandas assert method. @Ram-N , are you using vanilla Python as well? Not anaconda?

**Comment by Ram-N** [comment]
@hectormz I am using the anaconda's Python.
```
Python 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 22:01:29) 
[MSC v.1900 64 bit (AMD64)] :: Anaconda, Inc. on win32
```

**Comment by anzelpwj** [comment]
Would y'all like me to make the extra PR for the `expected = expected.astype({"null_flag": "int32"})` line?

**Comment by Ram-N** [comment]
@anzelpwj Yes, please. That will help me. I can rebase and continue.

**Pull request merged**